### PR TITLE
feat: sync webcomponent props with attributes

### PR DESCRIPTION
**Pull request by tpoisseau**
Refs: https://github.com/cheminfo/openchemlib-js/issues/229
Refs: https://github.com/cheminfo/openchemlib-js/pull/231

New PR, without copy-paste events handlings

Reminder: sync webcomponent properties with attributes, and propagate vaadin events ([need more discussions](https://github.com/cheminfo/openchemlib-js/issues/229#issuecomment-2401334225))

### CODE REVIEW

**Comment by tpoisseau** [comment]
The only modification I kept from original PR with copy-paste handling is to prevent to propagate events `ctrl+c` / `ctrl+v` to java event manager.

It will avoid to display errors in console (copy is not working outside a valid user-interaction in browser, at least for firefox).
And paste can't be implemented in sync.

**Comment by targos** [comment]
@artaius do you think it will work for you?

**Comment by artaius** [comment]
Thanks for the changes. As far as I can say this looks good to me.

The only missing thing I can see is, that the molecule coordinates are not yet considered on changes nor are they accessible by properties/attributes. If this is supported as well, we can start to deploy a first prototype to our users. The coords are crucial as people wont accept if the display of the molecules change upon pasting. 

Otherwise I found the solutions to
- set custom event names in Vaadin  (41fecf9)
- handle the clipboard directly though Vaadin's Element API.

**Comment by thsa** [comment]
For encoding reaction including mappinf coordinates, etc, you use the ReactionEncoder

**Comment by artaius** [comment]
> I think we can use the `"idcode idcoordinates"` convention for molecules, but what about reactions? it seems that the space character is already used to separate reactants from products?

Fully agree, lets keep this concatination of "idcode coords" for molecules. This way we also get change events on coord changes, what makes sense.

For reactions I think we can use the similar thing (as @thsa suggested). I think the following two methods should do. @thsa  Can you please confirm that this makes sense for the editor (especially the mode param)?

```java
// get reaction idcode
String reactionIdCode = ReactionEncoder.encode(new Reaction(), true, ReactionEncoder.INCLUDE_MAPPING | ReactionEncoder.INCLUDE_COORDS | ReactionEncoder.INCLUDE_DRAWING_OBJECTS | ReactionEncoder.INCLUDE_CATALYSTS | ReactionEncoder.RETAIN_REACTANT_AND_PRODUCT_ORDER);

// parse reaction idcode
Reaction decodedReaction = ReactionEncoder.decode(reactionIdCode, true);
```

**Comment by thsa** [comment]
It depends a little on the purpose. A typical case would be an encoding of a reaction with atom coords, mapping and original order of the reactants:
String reactionIdCode = ReactionEncoder.encode(new Reaction(), true, ReactionEncoder.INCLUDE_MAPPING | ReactionEncoder.INCLUDE_COORDS | ReactionEncoder.RETAIN_REACTANT_AND_PRODUCT_ORDER);

If catalysts structures are needed, the a typical UI would probably not allow drawing them within the reaction editor, but do it in a separate step.

If you need a canonical reaction representation, e.g. to remove duplicates from a database, then you would not include coords, mapping, catalysts. And you would allow the encoder to put reactants into a canonical order.

**Comment by artaius** [comment]
@thsa Thanks for the comments.

I would then go with @thsa proposal, i.e.
```java
// get reaction idcode
String reactionIdCode = ReactionEncoder.encode(new Reaction(), true, ReactionEncoder.INCLUDE_MAPPING | ReactionEncoder.INCLUDE_COORDS | ReactionEncoder.RETAIN_REACTANT_AND_PRODUCT_ORDER);

// parse reaction idcode
Reaction decodedReaction = ReactionEncoder.decode(reactionIdCode, true);
```

**Comment by tpoisseau** [comment]
I can use `ReactionEncoder.encode(this.getReaction(), true, ReactionEncoder.INCLUDE_MAPPING | ReactionEncoder.INCLUDE_COORDS | ReactionEncoder.RETAIN_REACTANT_AND_PRODUCT_ORDER);` / `ReactionEncoder.decode(reactionIdCode, true)` for reaction.

And `${molecule.getIDCode()} ${molecule.getIDCoordinates()}` / `const [idcode, coordinates] = this.idcode.split(' '); Molecule.fromIDCode(idcode, coordinates)`.

`ReactionEncoder` store some constants about separator it could use. Do we have constants to use for molecule separator ?

**Comment by thsa** [comment]
For testing and comparison to the Java version you could download openmolecules.org/idcodeexplorer.jar. It is an editor that has molecule and reaction modes, both for molecules and substructures. On the fly it creates idcodes/rxncodes with every editor change. You can also paste an idcode or rxncode and it converts it to the respective structures/reactions...

**Comment by artaius** [comment]
@tpoisseau Looks like if not all overloads of ```ReactionEncoder``` methods are present in the current OCL-JS.

- ```ReactionEncoder.decode(String)``` exists but ```ReactionEncoder.decode(String, boolean)``` doesn't.
- ```ReactionEncoder.encode(Reaction)``` exists but ```ReactionEncoder.encode(Reaction, boolean, int)``` doesn't.

File: src/com/actelion/research/gwt/core/JSReactionEncoder.java [hunk 1]
```diff
@@ -1,20 +1,87 @@
...
+  public static final int INCLUDE_DEFAULT = ReactionEncoder.INCLUDE_DEFAULT;
+
+  public static final String MOLECULE_DELIMITER = ReactionEncoder.MOLECULE_DELIMITER_STRING;
+  public static final String OBJECT_DELIMITER = ReactionEncoder.OBJECT_DELIMITER_STRING;
+  public static final String PRODUCT_IDENTIFIER = String.valueOf(ReactionEncoder.PRODUCT_IDENTIFIER);
+  public static final String CATALYST_DELIMITER = String.valueOf(ReactionEncoder.CATALYST_DELIMITER);
+
+  public static native String encode(JSReaction reaction, JavaScriptObject options) /*-{
+    if (!options) options = {keepAbsoluteCoordinates: false, mode: @com.actelion.research.gwt.core.JSReactionEncoder::INCLUDE_DEFAULT};
+    options.keepAbsoluteCoordinates = Boolean(options.keepAbsoluteCoordinates);
```
Comment (targos) [inline src/com/actelion/research/gwt/core/JSReactionEncoder.java:28]: don't modify the object passed by the user. same in other places below

File: __tests__/ReactionEncoder.test.js [hunk 2]
```diff
@@ -13,4 +13,44 @@ describe('ReactionEncoder class', () => {
...
+        ReactionEncoder.RETAIN_REACTANT_AND_PRODUCT_ORDER,
+    });
+    const idCodeToCompare = ReactionEncoder.encode(reaction, {
+      keepAbsoluteCoordinates: true,
+      mode:
+        ReactionEncoder.INCLUDE_DEFAULT |
+        ReactionEncoder.RETAIN_REACTANT_AND_PRODUCT_ORDER,
+    });
+
+    expect(idCodeToCompare).toBe(baseIdCode);
```
Comment (targos) [inline __tests__/ReactionEncoder.test.js:54]: I don't understand this test. You're doing exactly the same thing twice and expecting the same result. It's technically correct, but it's not really useful. Only the use of randomness in the tested functions could end up in a failed test.

Comment (tpoisseau) [reply __tests__/ReactionEncoder.test.js:54]: It was to ensure decode and encode with same options had same serialization. But yes, maybe it's not the responsibility of openchemlib-js to check that.  I'll remove the test

Comment (targos) [reply __tests__/ReactionEncoder.test.js]: Can't you instead test that the idcode generated by the JS API is the same that you had at the beginning (from the java program)?

**Comment by tpoisseau** [comment]
@artaius Where could I find `ID-Code` encoding options used by `idcodeexplorer.jar` ?

CleanShot 2024-10-17 at 09 13 29@2x
Give me 
```
gOp@DjWkjj`@ gOp@DjWkjj`@!gOp@DjWkjj`@##!BbOvw?_x@?g?~_{_} !Bb@Jw@oy??`C~@K\B !Bb@K~_{\BbOvH?_y?##
```

webcomponent transform it to 

```
gOp@DjWkjj`@ gOp@DjWkjj`@!gOp@DjWkjj`@##!RbGwW_Wx@_c}~O}]}GThB !RbOsW?Gx?_`A~@M_|GYkB !R|Gq~_{]||Gwp_Wy?GT[|##
```

Which look similar.

CleanShot 2024-10-17 at 09 17 22@2x

File: __tests__/ReactionEncoder.test.js (hunk 2 above)
Comment (tpoisseau) [reply __tests__/ReactionEncoder.test.js]: For that, I need to decode, then re-encode.
decode and encode have params, I don't know the right combinaisons to have a working tests.

Comment (tpoisseau) [reply __tests__/ReactionEncoder.test.js]: I added a parametrized ID-Code loader into https://229-improve-web-component-wi.openchemlib-js.pages.dev/examples/generic_editor/demo page.

I should help for functional testing

**Comment by artaius** [comment]
This is just a minor change of the coordinates. Not sure where this comes from. @thsa might this come from some floating point inaccuracy, or because of the different drawing area sizes.

@thsa Could you please give a hint about where to find the source code of the idcodeexplorer?
However I don't think this is a problem.

**Comment by artaius** [comment]
@tpoisseau Do you have any plans to release a (beta) version of the current code to npm? This would be great as I could then already adapt the Vaadin project to the new package and test it as well.

**Comment by tpoisseau** [comment]
As soon this PR is merged, we can release I guess, I'm not sure a beta is needed.

Changes concern mainly the webcomponent, and add options to ReactionEncoder encode-decode.

@artaius can you check the defaults params for ReactionEncoder to confirm the transition will be smooth please ?
https://github.com/cheminfo/openchemlib-js/pull/234/files#diff-a150fb53afee945f23f3fe7c176007111f3ba81bb0d188af4077d4a0fc4caa37

**Comment by thsa** [comment]
the source code of idcodeexplorer is not public, because it is meant as a test application for myself.

File: __tests__/ReactionEncoder.test.js (hunk 2 above)
Comment (tpoisseau) [reply __tests__/ReactionEncoder.test.js]: @artaius seems OK with results. to many actors interact to have stable idcode between editors (size of area, parameters of encode/decode, float precision)

So I close this conversation

**Comment by thsa** [comment]
the encoding part of the idcodeexplorer source is:
			if (mComboBoxMode.getSelectedIndex() == MODE_REACTION) {
				Reaction rxn = mDrawPanel.getDrawArea().getReaction();
				mTextFieldIDCode.setText(ReactionEncoder.encode(rxn, false,
						ReactionEncoder.INCLUDE_DEFAULT | ReactionEncoder.RETAIN_REACTANT_AND_PRODUCT_ORDER));
				mTextFieldSmiles.setText(rxn.isFragment() ?
					  IsomericSmilesCreator.createReactionSmiles(rxn)
					: IsomericSmilesCreator.createReactionSmarts(rxn));
				}
			else {
				StereoMolecule mol = new StereoMolecule(mMolecule);
				MoleculeStandardizer.standardize(mol,0);
				Canonizer canonizer = new Canonizer(mol, Canonizer.ENCODE_ATOM_CUSTOM_LABELS);
				mTextFieldIDCode.setText(canonizer.getIDCode() + " " + canonizer.getEncodedCoordinates());
				mTextFieldSmiles.setText(mMolecule.isFragment() ?
						IsomericSmilesCreator.createSmarts(mMolecule)
					  : IsomericSmilesCreator.createSmiles(mMolecule));
				}
			}

**Comment by tpoisseau** [comment]
ok, so you encode without absolute coordinates, So I guess it's normal import idcode from idcodeexplorer do not places the reactants exactly to same place when decode with `ensureCoordinates` to true.

**Comment by artaius** [comment]
> As soon this PR is merged, we can release I guess, I'm not sure a beta is needed.
> 
> Changes concern mainly the webcomponent, and add options to ReactionEncoder encode-decode.
> 
> @artaius can you check the defaults params for ReactionEncoder to confirm the transition will be smooth please ? https://github.com/cheminfo/openchemlib-js/pull/234/files#diff-a150fb53afee945f23f3fe7c176007111f3ba81bb0d188af4077d4a0fc4caa37

@tpoisseau As far as I can say this looks good to me... :+1:

**Comment by targos** [comment]
@tpoisseau Can you prepare a squash commit message and I'll merge?

**Comment by targos** [comment]
I just changed it to this for release-please:

```
feat: sync webcomponent props with attributes

Refs: https://github.com/cheminfo/openchemlib-js/issues/229

feat: add options to `ReactionEncoder.decode` and `ReactionEncoder.decode`
```

See https://github.com/googleapis/release-please#what-if-my-pr-contains-multiple-fixes-or-features

**Pull request merged**